### PR TITLE
feat: use mimalloc instead of musl's mallocng for static builds

### PR DESCRIPTION
**Pull request by dunglas**
The default musl allocator is slow for multi-threaded apps such as FrankenPHP and triggers weird bugs such as https://github.com/php/php-src/issues/13648.

The trickery in this patch (massively inspired by [this article by Tweag](https://www.tweag.io/blog/2023-08-10-rust-static-link-with-mimalloc/)) replaces the musl allocator by https://github.com/microsoft/mimalloc, a fast allocator created by Microsoft.

Hopefully, this will also help for https://github.com/dunglas/frankenphp/issues/481#issuecomment-1987287765.

### CODE REVIEW

**Comment by withinboredom** [comment]
Oho, gonna test this one real quick.

**Review by withinboredom — APPROVED** [review]
Appears to be ~150% faster

```
// libc:
[2024-03-15T17:58:12.419081+00:00] alert: Completed in 15.709 seconds  {"process_id":87753,"memory_peak_usage":"114 MB"}

// this patch
[2024-03-15T18:11:51.670770+00:00] alert: Completed in 40.284 seconds  {"process_id":62743,"memory_peak_usage":"110 MB"}

// vanilla musl
[2024-03-15T18:15:36.365410+00:00] alert: Completed in 51.751 seconds  {"process_id":4149,"memory_peak_usage":"110 MB"}
```

That being said, I did run into some issues with some caddy modules that needed to be disabled or I would get segfaults randomly.

**Comment by dunglas** [comment]
I also get random segfaults with the OpenSSL extension. This isn't ready for prime time 😅

File: build-static.sh [hunk 1]
```diff
@@ -117,28 +123,97 @@ if [ "${os}" = "mac" ]; then
     export CGO_LDFLAGS="-framework CoreFoundation -framework SystemConfiguration"
 fi
 
-CGO_LDFLAGS="${CGO_LDFLAGS} ${PWD}/buildroot/lib/libbrotlicommon.a ${PWD}/buildroot/lib/libbrotlienc.a ${PWD}/buildroot/lib/libbrotlidec.a $(./buildroot/bin/php-config --ldflags) $(./buildroot/bin/php-config --libs)"
+CGO_LDFLAGS="${CGO_LDFLAGS} ${PWD}/buildroot/lib/libbrotlicommon.a ${PWD}/buildroot/lib/libbrotlienc.a ${PWD}/buildroot/lib/libbrotlidec.a $(./buildroot/bin/php-config --ldflags || true) $(./buildroot/bin/php-config --libs || true)"
```
[the remark this answers is not in the document]
Comment (dunglas) [reply build-static.sh:126]: This is to prevent some shellcheck warnings.

**Comment by withinboredom** [comment]
Could it be the libraries are linked without mimalloc?

**Comment by dunglas** [comment]
@withinboredom could you give more details on the modules sefgfaulting in your tests? It looks like the OpenSSL issue may not be related to malloc after all.

**Comment by theofidry** [comment]
@dunglas why can't glibc be used there even if it means shipping something else than Alpine Linux? Or alternatively wouldn't it work with gcompat?

**Comment by dunglas** [comment]
@theofidry I'm not sure to follow. We provide glibc builds too (in the Docker images), but glibc doesn't support creating static builds. We have to use musl for that. See also: https://github.com/crazywhalecc/static-php-cli/issues/376

**Comment by andypost** [comment]
Just FYI lots of openssl tests fails in Alpinelinux with openssl (even years ago with libressl)

Current set of disabled openssl tests can be found at [repo](https://gitlab.alpinelinux.org/alpine/aports/-/blob/master/community/php83/disabled-tests.list#L31) 
At the end of the file new failures for # openssl 3.2.1
```
ext/openssl/tests/openssl_error_string_basic_openssl3.phpt
ext/openssl/tests/openssl_private_decrypt_basic.phpt
ext/openssl/tests/openssl_x509_parse_basic.phpt
```

**Comment by andypost** [comment]
Checked where alternate *malloc is used in Alpinelinux
- https://pkgs.alpinelinux.org/package/edge/main/x86_64/scudo-malloc (rust,firefox,lld)
- https://pkgs.alpinelinux.org/package/edge/main/x86_64/jemalloc (only blender)

**Comment by withinboredom** [comment]
I'll give this a test again this weekend.

**Comment by arnaud-lb** [comment]
@withinboredom what benchmark were you running in https://github.com/dunglas/frankenphp/pull/666#pullrequestreview-1939793858 ? In my [benchmarks](https://gist.github.com/arnaud-lb/5240aaa866ec6f662863dd5fe2bd10e9) the difference between glibc and musl are not as dramatic (and alternative allocators also don't do as much difference)

**Comment by withinboredom** [comment]
It's actually a benchmark of an app I'm building. It results in a lot of syscalls deep under the hood (lots of file access, shm, proxy generation, etc).

**Comment by dunglas** [comment]
I added a new experimental flag to build using mimalloc and added a CI pipeline to build a mimalloc variant automatically.

**Pull request merged**